File: problem004.py
# A palindromic number reads the same both ways.
# The largest palindrome made from the product of two -digit numbers is
# 9009 = 91 * 99
# Find the largest palindrome made from the product of two 3-digit numbers.


def is_palindrome(n):
    return str(n) == str(n)[::-1]


palindromes = [
    x * y
    for x in range(100, 999)
    for y in range(x, 999)
    if is_palindrome(x * y)
]

print(max(palindromes))

File: c/problem004.c
// A palindromic number reads the same both ways.
// The largest palindrome made from the product of two -digit numbers is
// 9009 = 91 * 99
// Find the largest palindrome made from the product of two 3-digit numbers.

#include <math.h>
#include <stdbool.h>
#include <stdio.h>

bool isPalindrome(int n) {
  int numDigits = log10(n) + 1;
  bool palindrome = true;

  for (int i = 0; i < numDigits / 2; i++) {
    palindrome &= ((n / (int)pow(10, i) % 10) ==
                   (n / (int)pow(10, numDigits - (i + 1)) % 10));
  }
  return palindrome;
}

int main() {
  int maxPal = 0;

  for (int x = 100; x < 999; x++) {
    for (int y = x; y < 999; y++) {
      maxPal = isPalindrome(x * y) && x * y > maxPal ? x * y : maxPal;
    }
  }
  printf("%d\n", maxPal);
  return 0;
}
